pyteomics/auxiliary/file_helpers.py

```diff
@@ -1121,26 +1134,206 @@ def iterate():
 
-        return iter(self._offset_index.keys())
+        Results will be returned out of order.
+
+        Parameters
+        ----------
+        target : :class:`Callable`, optional
+            The function to execute over each entry. It will be given a single object yielded by
+            the wrapped iterator as well as all of the values in ``args`` and ``kwargs``.
+
+            .. warning::
+                `target` must be thread-safe. The target function cannot interact with the underlying file object directly.
+
+        workers : int or None, optional
+            The number of worker threads to use. If not a positive integer,
+            defaults to the number of available CPUs.
+        args : :class:`Sequence`, optional
+            Additional positional arguments to be passed to the target function.
+        kwargs : :class:`Mapping`, optional
+            Additional keyword arguments to be passed to the target function.
+        **_kwargs
+            Additional keyword arguments to be passed to the target function.
+
+        Yields
+        ------
+        object
+            The work item returned by the target function.
+        """
+        if args is None:
+            args = tuple()
+        else:
+            args = tuple(args)
+        if kwargs is None:
+            kwargs = dict()
+        else:
+            kwargs = dict(kwargs)
+        kwargs.update(_kwargs)
+
+        # Determine number of workers
+        if workers is None or workers < 1:
+            workers = _NPROC
+
+        def worker_func(key, queue, local_reader):
+            try:
+                item = local_reader[key]
+                if target is not None:
+                    item = target(item, *args, **kwargs)
+                queue.put((True, item))
+            except Exception as e:
+                queue.put((False, e))
+
+        # Queues for distributing work and collecting results
+        output_queue = queue.Queue()
+        key_queue = queue.Queue()
+        worker_threads = []
+        _SENTINEL = object()
+
+        def producer():
+            """Iterate over task keys in a single thread and enqueue them for workers."""
+            try:
+                for key in self._task_map_iterator():
+                    key_queue.put(key)
+            finally:
+                # Signal completion to all worker threads
+                for _ in range(workers):
+                    key_queue.put(_SENTINEL)
+
+        def thread_worker():
+            # Initialize thread-local reader copy on first access
+            local_reader = copy(self)
+            try:
+                while True:
+                    key = key_queue.get()
+                    if key is _SENTINEL:
+                        # Do not put the sentinel back; each worker consumes one
+                        break
+                    worker_func(key, output_queue, local_reader)
```

Comment:
What is the value of not simply inlining `worker_func` here? This adds an extra function call overhead for no perceivable benefit as `worker_func` is a local function and not something that can be overridden.